viamonvif/service.go

```diff
@@ -131,14 +159,138 @@ func (dis *rtspDiscovery) DiscoverResources(ctx context.Context, extra map[strin
+
+		dataURL, err := downloadPreviewImage(ctx, dis.logger, snapshotURI)
+		if err != nil {
+			return nil, fmt.Errorf("failed to download preview image: %w", err)
+		}
+		return map[string]interface{}{
+			"preview": dataURL,
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("unknown command: %s", cmd)
+	}
+}
+
 func (dis *rtspDiscovery) Close(_ context.Context) error {
 	dis.mdnsServer.Shutdown()
 	return nil
 }
 
+func toPreviewCommand(command map[string]interface{}) (*previewRequest, error) {
+	attributes, ok := command["attributes"].(map[string]interface{})
+	if !ok {
+		return nil, errors.New("attributes is missing or not a map")
+	}
+	rtspURL, ok := attributes["rtsp_address"].(string)
+	if !ok {
+		return nil, errors.New("rtsp_address cannot be empty")
+	}
+	return &previewRequest{rtspURL: rtspURL}, nil
+}
+
+// formatDataURL formats the image data and content type into a data URL.
+func formatDataURL(contentType string, imageBytes []byte) string {
+	base64Image := base64.StdEncoding.EncodeToString(imageBytes)
+	return fmt.Sprintf("data:%s;base64,%s", contentType, base64Image)
+}
+
+// downloadPreviewImage downloads the preview image from the snapshot uri and returns it as a data URL.
+func downloadPreviewImage(ctx context.Context, logger logging.Logger, snapshotURI string) (string, error) {
+	parsedURL, err := url.Parse(snapshotURI)
+	if err != nil {
+		return "", fmt.Errorf("found an invalid snapshot URI: %w", err)
+	}
+
+	var username, password string
+	if parsedURL.User != nil {
+		username = parsedURL.User.Username()
+		if pwd, hasPassword := parsedURL.User.Password(); hasPassword {
+			password = pwd
+		}
+		if password == "" {
+			logger.Warnf("found a snapshot URI with no password: %s", snapshotURI)
+		}
+		logger.Debugf("creating snapshot request using credentials: username=%s", username)
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, //nolint:gosec
+	}
+	client := &http.Client{
+		// Setting upper bound timeout in case the ctx never times out
+		Timeout: httpClientTimeout,
+		Transport: &digest.Transport{
+			Username:  username,
+			Password:  password,
+			Transport: transport,
+		},
+	}
+
+	initialReq, err := http.NewRequestWithContext(ctx, http.MethodGet, snapshotURI, nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create http request: %w", err)
+	}
+
+	resp, err := client.Do(initialReq)
+	if err != nil {
+		return "", fmt.Errorf("failed to execute http request: %w", err)
+	}
+	defer resp.Body.Close()
+	logger.Debugf("snapshot response status: %s", resp.Status)
+
+	if resp.StatusCode != http.StatusOK {
+		statusText := http.StatusText(resp.StatusCode)
+		bodyPreview := "<could not read response body>"
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err == nil {
+			bodyPreview = string(bodyBytes)
+			// Truncate body preview if it's too long
+			if len(bodyPreview) > maxBodyStringSize {
+				bodyPreview = bodyPreview[:maxBodyStringSize] + "... [truncated]"
+			}
```

Comment:
We should be able to return the entire error, not matter how big. Please don't truncate the error message